### ISSUE TITLE
MongoDB health checks are raising user assertion errors.

### ISSUE THREAD

**Issue by richerlariviere** [kubedb/project#626]
I added some monitoring on a kubedb instance and noticed a weird constant increase in the user [assertion](https://docs.mongodb.com/manual/reference/command/serverStatus/#asserts) type (metric name: mongodb_asserts_total):

![Screen Shot 2019-08-22 at 1 49 57 PM](https://user-images.githubusercontent.com/5884148/63537441-be400e00-c4e3-11e9-94d8-11e77610a201.png)

I did some tests and found that the current form of health check is producing an assertion (error) each time it is invoked by Kubernetes. So I decided to exec into a mongodb replicaset (named coreos-prom-mgo) to see if the healthcheck is working:

```
livenessProbe:
  exec:
    command:
    - mongo
    - --eval
    - db.adminCommand('ping')
```
Becomes
```shell
$ kubectl exec -it coreos-prom-mgo-2 /bin/bash
root@coreos-prom-mgo-2:/$ mongo --eval db.adminCommand('ping')
bash: syntax error near unexpected token `('
```

But now if we add the double quotes:
```shell
root@coreos-prom-mgo-2:/$ mongo --eval "db.adminCommand('ping')"
MongoDB shell version v4.0.5
connecting to: mongodb://127.0.0.1:27017/?gssapiServiceName=mongodb
Implicit session: session { "id" : UUID("feabacc5-80f1-4123-a06c-548710d33de3") }
MongoDB server version: 4.0.5
{
        "ok" : 1,
        "operationTime" : Timestamp(1566496903, 1),
        "$clusterTime" : {
                "clusterTime" : Timestamp(1566496903, 1),
                "signature" : {
                        "hash" : BinData(0,"YB5v9ucra3ZOGw48YheQ/jbyHZU="),
                        "keyId" : NumberLong("6727977956066459650")
                }
        }
}
```
Therefore, `db.adminCommand('ping')` should be be surrounded by double quotes or else it will cause an assertion. Unfortunately, in Yaml, even if you do such thing, the parser will take them away.

So, I found a nice quick fix. Here is an example mongodb manifest:

```
apiVersion: kubedb.com/v1alpha1
kind: MongoDB
metadata:
  name: coreos-prom-mgo
  namespace: default
spec:
  version: "4.0.5-v1"
  terminationPolicy: WipeOut
  replicas: 3
  replicaSet:
    name: rs0
  podTemplate:
    spec:
      livenessProbe:
        exec:
          command:
          -  echo
          - -e
          - "--eval \"db.adminCommand('ping')\""
          -  \|
          -  xargs
          -  mongo
        failureThreshold: 3
        periodSeconds: 10
        successThreshold: 1
        timeoutSeconds: 5
      readinessProbe:
        exec:
          command:
          -  echo
          - -e
          - "--eval \"db.adminCommand('ping')\""
          -  \|
          -  xargs
          -  mongo
        failureThreshold: 3
        periodSeconds: 10
        successThreshold: 1
        timeoutSeconds: 1
  storage:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: 1Gi
  monitor:
    agent: prometheus.io/coreos-operator
    prometheus:
      namespace: default
      labels:
        release: name-of-your-prometheus-release-label-value
      interval: 10s
```

After doing that, the asserts errors stopped their constant increase.

Lesson of the day (to myself): Watch your MongoDB metrics folks it's really important!

**Comment by MrBones757** [comment]
Looks like a good pickup! I'm currently looking at doing a proof of concept around this tool so great fix as mongo is one of the dbs i'd be evaluating! :)

**Comment by the-redback** [comment]
I tried it and I can confirm that there is a continuous increase in `mongodb_asserts_total`. 

**There is a twist though**, When you exec into pod and run `mongo --eval "db.adminCommand('ping')"`, It may seem that it is passed but there is `Unauthorized` error in pod's mongod log. 

```
2019-08-23T05:36:23.423+0000 I NETWORK  [conn27] received client metadata from 127.0.0.1:59264 conn27: { application: { name: "MongoDB Shell" }, driver: { name: "MongoDB Internal Client", version: "3.6.12" }, os: { type: "Linux", name: "Ubuntu", architecture: "x86_64", version: "16.04" } }
2019-08-23T05:36:23.431+0000 I ACCESS   [conn27] Unauthorized: not authorized on admin to execute command { endSessions: [ { id: UUID("928633b8-a953-434f-b3bd-2a438464c4e5") } ], $db: "admin" }
2019-08-23T05:36:23.432+0000 I NETWORK  [conn27] end connection 127.0.0.1:59264 (2 connections now open)
``` 

And, this also increases the total amount of `mongodb_asserts_total`.  **so, `mongo --eval "db.adminCommand('ping')"` still generates [asserts.user](https://docs.mongodb.com/manual/reference/command/serverStatus/#serverstatus.asserts.user)**.

My guess, the user assertion is occuring whenever we connect without `auth`. So, `mongodb_asserts_total` is in increasing state because the `default probes` are connected without providing auth. https://github.com/kubedb/project/issues/609

If we replace probes with the following command, It passes successfully. Here bash is used to parse the evironment variables.
```
        exec:
          command:
          - bash
          - -c
          - |
            mongo --username="$MONGO_INITDB_ROOT_USERNAME" --password="$MONGO_INITDB_ROOT_PASSWORD" --eval "db.adminCommand('ping')"
```

So, the working yaml without increasing `mongodb_asserts_total`:

```
apiVersion: kubedb.com/v1alpha1
kind: MongoDB
metadata:
  name: mgo-replicaset
  namespace: demo
spec:
  certificateSecret:
    secretName: mgo-replicaset-keyfile
  databaseSecret:
    secretName: mgo-replicaset-auth
  monitor:
    agent: prometheus.io/coreos-operator
    prometheus:
      interval: 10s
      labels:
        k8s-app: prometheus
      namespace: monitoring
      port: 56790
    resources: {}
  podTemplate:
    controller: {}
    metadata: {}
    spec:
      livenessProbe:
        exec:
          command:
          - bash
          - -c
          - |
            mongo --username="$MONGO_INITDB_ROOT_USERNAME" --password="$MONGO_INITDB_ROOT_PASSWORD" --eval "db.adminCommand('ping')"
        failureThreshold: 3
        periodSeconds: 10
        successThreshold: 1
        timeoutSeconds: 5
      readinessProbe:
        exec:
          command:
          - bash
          - -c
          - |
            mongo --username="$MONGO_INITDB_ROOT_USERNAME" --password="$MONGO_INITDB_ROOT_PASSWORD" --eval "db.adminCommand('ping')"
        failureThreshold: 3
        periodSeconds: 10
        successThreshold: 1
        timeoutSeconds: 1
      resources: {}
  replicaSet:
    name: rs0
  replicas: 3
  serviceTemplate:
    metadata: {}
    spec: {}
  storage:
    accessModes:
    - ReadWriteOnce
    dataSource: null
    resources:
      requests:
        storage: 1Gi
    storageClassName: standard
  storageType: Durable
  terminationPolicy: Pause
  updateStrategy:
    type: RollingUpdate
  version: 3.6-v3

```

But, I am still wondering, 

- why the custom probes of @richerlariviere are not generating `unauthorized` warning in mongod log, where the defualt probes do? Both miss the authentication parameters.
- The default probe does not generate `unauthorized` log for `standalone` mongodb either.

**Comment by the-redback** [comment]
I found another Interesting issue related to mongodb ping. 

1. Even if ping fails, the command exit code is not set to `non-zero`. mongodb issue: [`mongo --eval` exits with status 0 on failure](https://jira.mongodb.org/browse/SERVER-38993)
  
```
root@mgo-replicaset-0:/# mongo --eval "db.adminCommand('ping')" --quiet
{
    "ok" : 1,
    "operationTime" : Timestamp(1566552534, 1),
    "$clusterTime" : {
        "clusterTime" : Timestamp(1566552534, 1),
        "signature" : {
            "hash" : BinData(0,"gtROXMOCrWm6MFCPUMFCXpI4AwU="),
            "keyId" : NumberLong("6728225827219046402")
        }
    }
}
root@mgo-replicaset-0:/# echo $?
0
root@mgo-replicaset-0:/# mongo --eval "db.adminCommand('ping2')" --quiet
{
    "operationTime" : Timestamp(1566552544, 1),
    "ok" : 0,
    "errmsg" : "no such command: 'ping2', bad cmd: '{ ping2: 1.0, lsid: { id: UUID(\"772b045c-96ec-4bbc-bd04-a44de99e4b85\") }, $clusterTime: { clusterTime: Timestamp(1566552544, 1), signature: { hash: BinData(0, E6A3DAAF7B64187D48437407C1FC49836AC6ACAA), keyId: 6728225827219046402 } }, $db: \"admin\" }'",
    "code" : 59,
    "codeName" : "CommandNotFound",
    "$clusterTime" : {
        "clusterTime" : Timestamp(1566552544, 1),
        "signature" : {
            "hash" : BinData(0,"5qPar3tkGH1IQ3QHwfxJg2rGrKo="),
            "keyId" : NumberLong("6728225827219046402")
        }
    }
}
root@mgo-replicaset-0:/# echo $?
0
```
Here, Exit status will only be non-zero, if `mongo` cli is unable to connect to a server. 

2. `--eval` is deprecated and will be unavailable after `4.2` release. https://docs.mongodb.com/manual/release-notes/4.2-compatibility/#remove-support-for-the-eval-command 


So, considering both, the below command can be a solution to go with. xref: https://stackoverflow.com/a/37852368/4628962

```
          command:
          - bash
          - -c
          - |
            if [[ $(echo 'db.runCommand("ping2").ok' | mongo localhost:27017 --username="$MONGO_INITDB_ROOT_USERNAME" --password="$MONGO_INITDB_ROOT_PASSWORD" --authenticationDatabase=admin --quiet ) -ne "1" ]]; then 
              exit 1
            fi
```

**Comment by richerlariviere** [comment]
@the-redback Thank you so much for your response and contribution on this issue. I gathered some details if it can help.

>     * why the custom probes of @richerlariviere are not generating `unauthorized` warning in mongod log, where the default probes do? Both miss the authentication parameters.

My probe is generating these messages each time a new connection is made.
```
I NETWORK  [listener] connection accepted from 127.0.0.1:44718 #6948 (6 connections now open)
I ACCESS   [conn6948] Successfully authenticated as principal root on admin 
```

>     * The default probe does not generate `unauthorized` log for `standalone` mongodb either.
If by standalone you mean a mongodb with only 1 replica, I have seen that behavior in one of my MongoDBs. I noticed an increase even if I'm not seing `unauthorized` warnings in my logs.

**Graph with 1 replica (with my custom probes): Increase**
Kubernetes version: 1.13.9
Node count: 3 (this is our staging cluster for internal testings)
Kubedb: 0.12.0 with `monitoring.enabled=true` and `monitoring.agent="prometheus.io/coreos-operator"`

<details>
  <summary>yaml manifest</summary>


</details>

![Screen Shot 2019-08-23 at 8 59 57 AM](https://user-images.githubusercontent.com/5884148/63594426-67d6dc00-c584-11e9-84c0-32a429c1154b.png)

**Graph with 3 replicas (with my custom probes): Pretty stable. That replicaset uses the same manifest I posted earlier.**
Setup:
Kubernetes version: 1.14.6
Node count: 1 (this is my test cluster)
Kubedb: 0.12.0 with `monitoring.enabled=true` and `monitoring.agent="prometheus.io/coreos-operator"`
![Screen Shot 2019-08-23 at 8 54 34 AM](https://user-images.githubusercontent.com/5884148/63594490-8d63e580-c584-11e9-8713-d7264cda75dd.png)

**Comment by richerlariviere** [comment]
I found something very interesting in my 1 replica setup in the exporter logs:

```
E0823 13:25:56.261459       1 replset_status.go:232] Failed to get replSet status.
E0823 13:25:56.261967       1 oplog_status.go:127] Failed to get local.oplog_rs collection stats. 
```

**Comment by the-redback** [comment]
For future debug about assertion error, increate log level: 

```console 
$ mongo --host=localhost --username="$MONGO_INITDB_ROOT_USERNAME" --password="$MONGO_INITDB_ROOT_PASSWORD" --authenticationDatabase=admin
> db.setLogLevel(10)
```

**Comment by the-redback** [comment]
Write now, two user assertion is occuring. one for unauthorized probe and another for exporter.

TODO:

- [ ] Fix unauthorized readiness and liveness probe
- [ ] Fix exporter caused user assertion error. [exporter is trying to use local db and it is causing user assertion in mongos]

**Comment by the-redback** [comment]
closing the issue as New exporter is introduced and also user assertion due to readiness and liveness probe is fixed. 

The fixes will be part of 0.13.0 release.

**Comment by richerlariviere** [comment]
Thank you so much!
